alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/navigationstrategies/RouteFollowing.kt

```diff
@@ -0,0 +1,89 @@
+import it.unibo.alchemist.model.implementations.utils.origin
+import it.unibo.alchemist.model.interfaces.EuclideanNavigationAction
+import it.unibo.alchemist.model.interfaces.NavigationStrategy
+import it.unibo.alchemist.model.interfaces.geometry.euclidean.twod.ConvexPolygon
+import it.unibo.alchemist.model.interfaces.geometry.euclidean.twod.Euclidean2DConvexShape
+import it.unibo.alchemist.model.interfaces.graph.Euclidean2DPassage
+
+/**
+ * A [NavigationStrategy] allowing to follow a given [route].
+ * The [route] consists of a list of positions (= waypoints) that may or may not be in sight of each
+ * other (i.e. the path leading from a waypoint to the next one may or may not be representable
+ * as a single segment), for this reason [Pursuing] behavior is used to reach each waypoint.
+ * In this context, a waypoint is considered reached when it's inside the current room (not when the
+ * pedestrian reach that exact position), apart from the last waypoint which is actually approached.
+ * Cuts to the route are allowed (i.e. if the pedestrian finds a waypoint which is farther than the
+ * expected next one, he/she skips all the waypoints in between).
+ *
+ * @param T the concentration type.
+ * @param N the type of landmarks of the pedestrian's cognitive map.
+ * @param E the type of edges of the pedestrian's cognitive map.
+ */
+open class RouteFollowing<T, N : Euclidean2DConvexShape, E> constructor(
+    action: EuclideanNavigationAction<T, N, E, ConvexPolygon, Euclidean2DPassage>,
+    /**
+     * The route to follow.
+     */
+    protected open val route: List<Euclidean2DPosition>
+) : DynamicPursuing<T, N, E>(action, action.environment.origin().takeIf { route.isEmpty() } ?: route[0]) {
+
+    private var indexOfNextWaypoint: Int = 0
+
+    /**
+     * When in an unexpected room the pedestrian gets back to [previousRoom] so as to continue following the
+     * route correctly.
+     */
+    override fun inUnexpectedNewRoom(
+        previousRoom: ConvexPolygon,
+        expectedNewRoom: ConvexPolygon,
+        actualNewRoom: ConvexPolygon
+    ) = with(action) {
+        doorsInSight()
+            .filter { it.head == previousRoom }
+            .minBy { it.distanceToPedestrian() }
+            ?.let { crossDoor(it) }
+            ?: inNewRoom(actualNewRoom)
+    }
+
+    override fun inNewRoom(newRoom: ConvexPolygon) {
+        if (route.isEmpty()) {
+            action.stop()
+        } else {
+            if (waypointReached(newRoom)) {
+                setDestination(route[indexOfNextWaypoint])
+            }
+            super.inNewRoom(newRoom)
+        }
+    }
+
+    /**
+     * Checks if a new waypoint has been reached and updates [indexOfNextWaypoint] accordingly.
+     */
+    protected fun waypointReached(currentRoom: ConvexPolygon): Boolean = false
+        .takeIf { indexOfNextWaypoint >= route.size - 1 }
+        ?: route.indexOfLast { isReached(it, currentRoom) }.let { indexOfFartherWaypointReached ->
```

Comment:
let on a new line, plus I think this can get rewritten more cleanly using `also` to encapsulate the side effect

Reply:
I rewrote it using more variables instead of `also`:
```kotlin
if (indexOfNextWaypoint >= route.size - 1) {
    return false
}
val indexOfFartherWaypointReached = route.indexOfLast { isReached(it, currentRoom) }
val newWaypointReached = indexOfFartherWaypointReached >= indexOfNextWaypoint
if (newWaypointReached) {
    indexOfNextWaypoint = // update it
}
return newWaypointReached

```
 I think this is the most readable way of writing it.